enumerables.rb

```diff
@@ -0,0 +1,153 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    incr = 0
+    arr_kind = self.class
+    if arr_kind == Hash
+      arr = flatten
+      incr = 2
+    else
+      arr = to_a
+      incr = 1
+    end
+
+    i = 0
+    while i < arr.length
+      arr_kind == Hash ? yield(arr[i], arr[i + 1]) : yield(arr[i])
+      i += incr
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = to_a
+
+    for i in 0...arr.length
+      yield(arr[i], i)
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    kind = self.class
+    arr = to_a
+
+    if kind == Hash
+      match = {}
+      flat = arr.flatten
+      i = 0
+      while i < flat.length
+        found = yield(flat[i], flat[i + 1])
+        match[flat[i]] = flat[i + 1] if found
+        i += 2
+      end
+    else
+      match = []
+      arr.my_each do |val|
+        found = yield(val)
+        match << val if found
+      end
+    end
+    match
+  end
+
+  def my_all?(pattern = nil)
```

Comment:
This method is good but you are missing some functionality of the original method:
- It should handle when a class is passed
- It should handle when a pattern like a regular expression is passed, which you attempted but please remember that you should check if block is given first and also consider another approach of testing your pattern
As you can see from `Class` requirement pattern argument can also be a class or even a Regexp or another pattern other than Class or Regular Expression so you need to check if it is a `Class` or find a way that works whether it's a Class or Regular Exp or another pattern

Read the [documentation](https://ruby-doc.org/core-2.7.0/Enumerable.html#method-i-all-3F) on `all?` for more info

Reply:
I added all the functionality requested, however, I'm not sure if there're patterns other than Regexp, Classes, and Simple objects that this should handle.